### PR TITLE
Add `Arena::retain` method

### PR DESCRIPTION
**Pull request by sdleffler**

### CODE REVIEW

File: src/arena.rs [hunk 1]
```diff
@@ -288,6 +288,34 @@ impl<T> Arena<T> {
...
+                    *entry = Entry::Empty(EmptyEntry {
+                        generation: occupied.generation,
+                        next_free: self.first_free,
+                    });
+
+                    // The next time we insert, we can re-use the empty entry we
+                    // just created. If another removal happens before then, that
+                    // entry will be used before this one (FILO).
+                    self.first_free = Some(FreePointer::from_slot(index.slot));
+                    self.len = self.len.checked_sub(1).unwrap_or_else(|| unreachable!());
```
Comment (LPGhatguy) [inline src/arena.rs:314]: We know fairly trivially that this len will never underflow -- we just verified that we have an item in the map. Can you add a comment pointing that out for future adventurers?

Comment (sdleffler) [reply src/arena.rs:314]: Sure; I actually copied that directly from `remove`, which is why that comment is there. I'll change the comment to reflect its new purpose.

Comment (LPGhatguy) [reply src/arena.rs:314]: Ahah, maybe we should add a comment to both cases! I think the comment about `first_free` is good, mostly just making a new block comment for the `self.len` assignment.

Comment (sdleffler) [reply src/arena.rs:317]: Sure! That should be done as a different PR though I think. Comment updated here, anyhow.

**Review by LPGhatguy — APPROVED** [review]

**Pull request merged**